File: engines/go/ai/expectimax.go
package ai

import (
	"encoding/json"
	"fmt"
	"math"
	"time"
)

// Config is the expectimax configuration (AI.md §3).
type Config struct {
	// AutoDepth selects depth = clamp(distinct-2, MinDepth, MaxDepth); otherwise Depth is fixed.
	AutoDepth        bool
	Depth            int
	MinDepth         int
	MaxDepth         int
	FourPruneEmpties int
	// TimeBudgetMs > 0 enables iterative deepening (non-deterministic).
	TimeBudgetMs float64
	// TTBits is log2 of the transposition-table slot count.
	TTBits  int
	Weights Weights
}

// CanonicalConfig returns the canonical profile.
func CanonicalConfig() Config {
	return Config{AutoDepth: true, MinDepth: 2, MaxDepth: 4, TTBits: 20, Weights: HeuristicV1}
}

// MarshalJSON renders depth as "auto" or an integer, matching the TS descriptor.
func (c Config) MarshalJSON() ([]byte, error) {
	var depth any = c.Depth
	if c.AutoDepth {
		depth = "auto"
	}
	return json.Marshal(struct {
		Depth            any     `json:"depth"`
		MinDepth         int     `json:"minDepth"`
		MaxDepth         int     `json:"maxDepth"`
		FourPruneEmpties int     `json:"fourPruneEmpties"`
		TimeBudgetMs     float64 `json:"timeBudgetMs"`
		TTBits           int     `json:"ttBits"`
		Weights          Weights `json:"weights"`
	}{depth, c.MinDepth, c.MaxDepth, c.FourPruneEmpties, c.TimeBudgetMs, c.TTBits, c.Weights})
}

func asInt(key string, v any) (int64, error) {
	f, ok := v.(float64)
	if !ok {
		if i, ok2 := v.(int); ok2 {
			return int64(i), nil
		}
		return 0, fmt.Errorf("config %s must be a number", key)
	}
	if f != math.Trunc(f) || math.IsInf(f, 0) {
		return 0, fmt.Errorf("config %s must be an integer (got %v)", key, f)
	}
	return int64(f), nil
}

// ParseConfig overlays a JSON-decoded config object onto the canonical profile.
func ParseConfig(m map[string]any) (Config, error) {
	c := CanonicalConfig()
	for k, v := range m {
		var err error
		var n int64
		switch k {
		case "depth":
			if s, ok := v.(string); ok {
				if s != "auto" {
					return c, fmt.Errorf("config depth must be an integer or \"auto\"")
				}
				c.AutoDepth = true
			} else {
				n, err = asInt(k, v)
				c.AutoDepth, c.Depth = false, int(n)
				if err == nil && n < 1 {
					err = fmt.Errorf("config depth must be >= 1")
				}
			}
		case "minDepth":
			n, err = asInt(k, v)
			c.MinDepth = int(n)
		case "maxDepth":
			n, err = asInt(k, v)
			c.MaxDepth = int(n)
		case "fourPruneEmpties":
			n, err = asInt(k, v)
			c.FourPruneEmpties = int(n)
		case "timeBudgetMs":
			f, ok := v.(float64)
			if !ok {
				err = fmt.Errorf("config timeBudgetMs must be a number")
			}
			c.TimeBudgetMs = f
		case "ttBits":
			n, err = asInt(k, v)
			if err == nil && (n < 4 || n > 28) {
				err = fmt.Errorf("config ttBits out of range")
			}
			c.TTBits = int(n)
		case "weights":
			wm, ok := v.(map[string]any)
			if !ok {
				return c, fmt.Errorf("config weights must be an object")
			}
			for wk, wv := range wm {
				n, err := asInt("weights."+wk, wv)
				if err != nil {
					return c, fmt.Errorf("heuristic weight '%s' must be an integer (got %v)", wk, wv)
				}
				switch wk {
				case "lost":
					c.Weights.Lost = n
				case "empty":
					c.Weights.Empty = n
				case "merges":
					c.Weights.Merges = n
				case "mono":
					c.Weights.Mono = n
				case "sum":
					c.Weights.Sum = n
				case "smooth":
					c.Weights.Smooth = n
				case "stable":
					c.Weights.Stable = n
				case "corner":
					c.Weights.Corner = n
				}
			}
		}
		if err != nil {
			return c, err
		}
	}
	return c, nil
}

// SearchResult is the outcome of one search.
type SearchResult struct {
	Move            int // -1 if no valid move
	Value           float64
	Values          [4]*float64
	Depth           int
	Nodes           int64
	TTHits          int64
	TTSize          int
	TimeUs          int64
	Deterministic   bool
	CompletedDepths []int
}

// transposition table: exact key (board, depth). Mirrors the TS reference
// (hash, 4-slot linear probe, store policy, clear at 75%) so that node and
// hit counts are comparable across ports.
type ttable struct {
	mask  uint32
	board []uint64
	depth []uint8
	value []float64
	size  int
}

func newTT(bits int) *ttable {
	n := 1 << bits
	return &ttable{mask: uint32(n - 1), board: make([]uint64, n), depth: make([]uint8, n), value: make([]float64, n)}
}

func (t *ttable) slot(b uint64, d int) uint32 {
	lo, hi := uint32(b), uint32(b>>32)
	h := ((lo ^ uint32(d)*0x9e3779b1) * 0x85ebca6b) ^ (hi * 0xc2b2ae35)
	h ^= h >> 15
	h *= 0x2c1b3c6d
	h ^= h >> 12
	return h & t.mask
}

func (t *ttable) clear() {
	clear(t.depth)
	t.size = 0
}

// Search is a reusable expectimax searcher (not safe for concurrent use).
type Search struct {
	Config   Config
	table    *[65536]float64
	corner   int64
	tt       *ttable
	nodes    int64
	ttHits   int64
	deadline time.Time
	timed    bool
	aborted  bool
}

// NewSearch creates a searcher with a preallocated transposition table.
func NewSearch(c Config) *Search {
	if c.TTBits == 0 {
		c.TTBits = 20
	}
	return &Search{Config: c, table: LineTable(c.Weights), corner: c.Weights.Corner, tt: newTT(c.TTBits)}
}

// Evaluate is the heuristic under this searcher's weights.
func (s *Search) Evaluate(b Bitboard) float64 { return Evaluate(b, s.table, s.corner) }

// DepthFor returns the search depth for b.
func (s *Search) DepthFor(b Bitboard) int {
	c := &s.Config
	if !c.AutoDepth {
		return c.Depth
	}
	return max(c.MinDepth, min(c.MaxDepth, DistinctRanks(b)-2))
}

func (s *Search) maxnode(b Bitboard, d int) float64 {
	s.nodes++
	if d == 0 {
		return Evaluate(b, s.table, s.corner)
	}
	best := 0.0
	for dir := 0; dir < 4; dir++ {
		nb, ok := Move(b, dir)
		if !ok {
			continue
		}
		v := s.chance(nb, d)
		if s.aborted {
			return 0
		}
		if v > best {
			best = v
		}
	}
	return best
}

func (s *Search) chance(b Bitboard, d int) float64 {
	s.nodes++
	if s.timed && s.nodes&0xfff == 0 && time.Now().After(s.deadline) {
		s.aborted = true
		return 0
	}
	tt := s.tt
	home := tt.slot(b, d)
	sl := home
	for probe := 0; probe < 4; probe++ {
		sd := tt.depth[sl]
		if sd == 0 {
			break
		}
		if int(sd) == d && tt.board[sl] == b {
			s.ttHits++
			return tt.value[sl]
		}
		sl = (sl + 1) & tt.mask
	}

	n := 0
	for i := 0; i < 16; i++ {
		if b>>(4*i)&0xf == 0 {
			n++
		}
	}
	p := s.Config.FourPruneEmpties
	four := !(p > 0 && n >= p)
	sum := 0.0
	for i := 0; i < 16; i++ {
		sh := uint(4 * i)
		if b>>sh&0xf != 0 {
			continue
		}
		if four {
			v2 := s.maxnode(b|1<<sh, d-1)
			if s.aborted {
				return 0
			}
			sum = sum + float64(0.9*v2)
			v4 := s.maxnode(b|2<<sh, d-1)
			if s.aborted {
				return 0
			}
			sum = sum + float64(0.1*v4)
		} else {
			v2 := s.maxnode(b|1<<sh, d-1)
			if s.aborted {
				return 0
			}
			sum = sum + v2
		}
	}
	v := sum / float64(n)

	target := home
	sl = home
	for probe := 0; probe < 4; probe++ {
		if tt.depth[sl] == 0 {
			target = sl
			tt.size++
			break
		}
		sl = (sl + 1) & tt.mask
	}
	tt.board[target] = b
	tt.depth[target] = uint8(d)
	tt.value[target] = v
	return v
}

type rootResult struct {
	move   int
	value  float64
	values [4]*float64
}

func (s *Search) root(b Bitboard, depth int) (rootResult, bool) {
	r := rootResult{move: -1, value: math.Inf(-1)}
	for dir := 0; dir < 4; dir++ {
		nb, ok := Move(b, dir)
		if !ok {
			continue
		}
		v := s.chance(nb, depth)
		if s.aborted {
			return r, false
		}
		vv := v
		r.values[dir] = &vv
		if v > r.value {
			r.value, r.move = v, dir
		}
	}
	return r, true
}

// Run searches b and returns the decision.
func (s *Search) Run(b Bitboard) SearchResult {
	start := time.Now()
	s.nodes, s.ttHits = 0, 0
	s.aborted, s.timed = false, false
	if float64(s.tt.size) > float64(s.tt.mask+1)*0.75 {
		s.tt.clear()
	}
	if s.Config.TimeBudgetMs <= 0 {
		depth := s.DepthFor(b)
		r, _ := s.root(b, depth)
		return s.result(r, depth, start, true, []int{depth})
	}
	best, _ := s.root(b, 1)
	depth := 1
	completed := []int{1}
	s.timed = true
	s.deadline = start.Add(time.Duration(s.Config.TimeBudgetMs * float64(time.Millisecond)))
	for d := 2; d <= s.Config.MaxDepth; d++ {
		r, ok := s.root(b, d)
		if !ok {
			break
		}
		best, depth = r, d
		completed = append(completed, d)
	}
	s.timed, s.aborted = false, false
	return s.result(best, depth, start, false, completed)
}

func (s *Search) result(r rootResult, depth int, start time.Time, det bool, completed []int) SearchResult {
	return SearchResult{
		Move: r.move, Value: r.value, Values: r.values, Depth: depth,
		Nodes: s.nodes, TTHits: s.ttHits, TTSize: s.tt.size,
		TimeUs:        roundUs(time.Since(start)),
		Deterministic: det, CompletedDepths: completed,
	}
}

func roundUs(d time.Duration) int64 { return int64(math.Round(float64(d.Nanoseconds()) / 1000)) }

File: engines/csharp/src/Ai/Expectimax.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace G2048.Ai;

/// <summary>Expectimax configuration (AI.md §3).</summary>
public sealed class Config
{
    /// <summary>depth = clamp(distinct-2, MinDepth, MaxDepth) when set; otherwise Depth is fixed.</summary>
    public bool AutoDepth { get; set; } = true;
    public int Depth { get; set; }
    public int MinDepth { get; set; } = 2;
    public int MaxDepth { get; set; } = 4;
    public int FourPruneEmpties { get; set; }
    /// <summary>&gt; 0 enables iterative deepening (non-deterministic).</summary>
    public double TimeBudgetMs { get; set; }
    /// <summary>log2 of the transposition-table slot count.</summary>
    public int TTBits { get; set; } = 20;
    public Weights Weights { get; set; } = Weights.V1;

    public static Config Canonical() => new();

    private static bool TryNumber(JsonNode? v, out double d)
    {
        d = 0;
        return v is JsonValue jv && jv.GetValueKind() == JsonValueKind.Number
            && double.TryParse(jv.ToJsonString(), NumberStyles.Float, CultureInfo.InvariantCulture, out d);
    }

    private static long AsInt(string key, JsonNode? v)
    {
        if (!TryNumber(v, out double f)) throw new ArgumentException($"config {key} must be a number");
        if (f != Math.Truncate(f) || double.IsInfinity(f)) throw new ArgumentException($"config {key} must be an integer (got {f})");
        return (long)f;
    }

    /// <summary>Overlays a JSON config object onto the canonical profile; throws ArgumentException.</summary>
    public static Config Parse(JsonObject? m)
    {
        var c = Canonical();
        if (m is null) return c;
        foreach (var (k, v) in m)
        {
            switch (k)
            {
                case "depth":
                    if (v is JsonValue sv && sv.GetValueKind() == JsonValueKind.String)
                    {
                        if (sv.GetValue<string>() != "auto") throw new ArgumentException("config depth must be an integer or \"auto\"");
                        c.AutoDepth = true;
                    }
                    else
                    {
                        long n = AsInt(k, v);
                        if (n < 1) throw new ArgumentException("config depth must be >= 1");
                        c.AutoDepth = false;
                        c.Depth = (int)n;
                    }
                    break;
                case "minDepth": c.MinDepth = (int)AsInt(k, v); break;
                case "maxDepth": c.MaxDepth = (int)AsInt(k, v); break;
                case "fourPruneEmpties": c.FourPruneEmpties = (int)AsInt(k, v); break;
                case "timeBudgetMs":
                    if (!TryNumber(v, out double tb)) throw new ArgumentException("config timeBudgetMs must be a number");
                    c.TimeBudgetMs = tb;
                    break;
                case "ttBits":
                {
                    long n = AsInt(k, v);
                    if (n < 4 || n > 28) throw new ArgumentException("config ttBits out of range");
                    c.TTBits = (int)n;
                    break;
                }
                case "weights":
                {
                    if (v is not JsonObject wm) throw new ArgumentException("config weights must be an object");
                    var w = c.Weights;
                    foreach (var (wk, wv) in wm)
                    {
                        long n;
                        try { n = AsInt("weights." + wk, wv); }
                        catch (ArgumentException) { throw new ArgumentException($"heuristic weight '{wk}' must be an integer (got {wv?.ToJsonString()})"); }
                        w = wk switch
                        {
                            "lost" => w with { Lost = n },
                            "empty" => w with { Empty = n },
                            "merges" => w with { Merges = n },
                            "mono" => w with { Mono = n },
                            "sum" => w with { Sum = n },
                            "smooth" => w with { Smooth = n },
                            "stable" => w with { Stable = n },
                            "corner" => w with { Corner = n },
                            _ => w,
                        };
                    }
                    c.Weights = w;
                    break;
                }
            }
        }
        return c;
    }
}

/// <summary>Outcome of one search.</summary>
public sealed class SearchResult
{
    /// <summary>-1 if no valid move.</summary>
    public int Move { get; init; } = -1;
    public double Value { get; init; }
    public double?[] Values { get; init; } = new double?[4];
    public int Depth { get; init; }
    public long Nodes { get; init; }
    public long TTHits { get; init; }
    public int TTSize { get; init; }
    public long TimeUs { get; init; }
    public bool Deterministic { get; init; }
    public int[] CompletedDepths { get; init; } = [];
}

/// <summary>Reusable expectimax searcher (not thread-safe).</summary>
public sealed class Search
{
    public Config Config { get; }

    private readonly double[] _table;
    private readonly long _corner;

    // Transposition table: exact key (board, depth). Mirrors the TS reference
    // (hash, 4-slot linear probe, store policy, clear at 75%) so node and hit
    // counts are comparable across ports.
    private readonly uint _ttMask;
    private readonly ulong[] _ttBoard;
    private readonly byte[] _ttDepth;
    private readonly double[] _ttValue;
    private int _ttSize;

    private long _nodes;
    private long _ttHits;
    private long _deadline;
    private bool _timed;
    private bool _aborted;

    public Search(Config c)
    {
        if (c.TTBits == 0) c.TTBits = 20;
        Config = c;
        _table = Heuristic.LineTable(c.Weights);
        _corner = c.Weights.Corner;
        int n = 1 << c.TTBits;
        _ttMask = (uint)(n - 1);
        _ttBoard = new ulong[n];
        _ttDepth = new byte[n];
        _ttValue = new double[n];
    }

    public double Evaluate(ulong b) => Heuristic.Evaluate(b, _table, _corner);

    public int DepthFor(ulong b)
    {
        var c = Config;
        if (!c.AutoDepth) return c.Depth;
        return Math.Max(c.MinDepth, Math.Min(c.MaxDepth, Bitboard.DistinctRanks(b) - 2));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private uint Slot(ulong b, int d)
    {
        unchecked
        {
            uint lo = (uint)b, hi = (uint)(b >> 32);
            uint h = ((lo ^ ((uint)d * 0x9e3779b1)) * 0x85ebca6b) ^ (hi * 0xc2b2ae35);
            h ^= h >> 15;
            h *= 0x2c1b3c6d;
            h ^= h >> 12;
            return h & _ttMask;
        }
    }

    private double MaxNode(ulong b, int d)
    {
        _nodes++;
        if (d == 0) return Heuristic.Evaluate(b, _table, _corner);
        double best = 0.0;
        for (int dir = 0; dir < 4; dir++)
        {
            ulong nb = Bitboard.Move(b, dir);
            if (nb == b) continue;
            double v = Chance(nb, d);
            if (_aborted) return 0;
            if (v > best) best = v;
        }
        return best;
    }

    private double Chance(ulong b, int d)
    {
        _nodes++;
        if (_timed && (_nodes & 0xfff) == 0 && Stopwatch.GetTimestamp() > _deadline)
        {
            _aborted = true;
            return 0;
        }
        ref ulong ttBoard = ref MemoryMarshal.GetArrayDataReference(_ttBoard);
        ref byte ttDepth = ref MemoryMarshal.GetArrayDataReference(_ttDepth);
        ref double ttValue = ref MemoryMarshal.GetArrayDataReference(_ttValue);
        uint mask = _ttMask;
        uint home = Slot(b, d);
        uint sl = home;
        for (int probe = 0; probe < 4; probe++)
        {
            byte sd = Unsafe.Add(ref ttDepth, (nint)sl);
            if (sd == 0) break;
            if (sd == d && Unsafe.Add(ref ttBoard, (nint)sl) == b)
            {
                _ttHits++;
                return Unsafe.Add(ref ttValue, (nint)sl);
            }
            sl = (sl + 1) & mask;
        }

        int n = Bitboard.CountEmpty(b);
        int p = Config.FourPruneEmpties;
        bool four = !(p > 0 && n >= p);
        double sum = 0.0;
        for (int i = 0; i < 16; i++)
        {
            int sh = 4 * i;
            if (((b >> sh) & 0xf) != 0) continue;
            if (four)
            {
                double v2 = MaxNode(b | (1UL << sh), d - 1);
                if (_aborted) return 0;
                // explicit product then add: RyuJIT never contracts this into FMA
                double p2 = 0.9 * v2;
                sum += p2;
                double v4 = MaxNode(b | (2UL << sh), d - 1);
                if (_aborted) return 0;
                double p4 = 0.1 * v4;
                sum += p4;
            }
            else
            {
                double v2 = MaxNode(b | (1UL << sh), d - 1);
                if (_aborted) return 0;
                sum += v2;
            }
        }
        double v = sum / n;

        uint target = home;
        sl = home;
        for (int probe = 0; probe < 4; probe++)
        {
            if (Unsafe.Add(ref ttDepth, (nint)sl) == 0)
            {
                target = sl;
                _ttSize++;
                break;
            }
            sl = (sl + 1) & mask;
        }
        Unsafe.Add(ref ttBoard, (nint)target) = b;
        Unsafe.Add(ref ttDepth, (nint)target) = (byte)d;
        Unsafe.Add(ref ttValue, (nint)target) = v;
        return v;
    }

    private (int Move, double Value, double?[] Values, bool Ok) Root(ulong b, int depth)
    {
        int move = -1;
        double value = double.NegativeInfinity;
        var values = new double?[4];
        for (int dir = 0; dir < 4; dir++)
        {
            ulong nb = Bitboard.Move(b, dir);
            if (nb == b) continue;
            double v = Chance(nb, depth);
            if (_aborted) return (move, value, values, false);
            values[dir] = v;
            if (v > value)
            {
                value = v;
                move = dir;
            }
        }
        return (move, value, values, true);
    }

    /// <summary>Searches b and returns the decision.</summary>
    public SearchResult Run(ulong b)
    {
        long start = Stopwatch.GetTimestamp();
        _nodes = 0;
        _ttHits = 0;
        _aborted = false;
        _timed = false;
        if (_ttSize > (_ttMask + 1.0) * 0.75)
        {
            Array.Clear(_ttDepth);
            _ttSize = 0;
        }
        if (Config.TimeBudgetMs <= 0)
        {
            int depth = DepthFor(b);
            var r = Root(b, depth);
            return Result(r.Move, r.Value, r.Values, depth, start, true, [depth]);
        }
        var best = Root(b, 1);
        int bestDepth = 1;
        var completed = new List<int> { 1 };
        _timed = true;
        _deadline = start + (long)(Config.TimeBudgetMs * Stopwatch.Frequency / 1000.0);
        for (int d = 2; d <= Config.MaxDepth; d++)
        {
            var r = Root(b, d);
            if (!r.Ok) break;
            best = r;
            bestDepth = d;
            completed.Add(d);
        }
        _timed = false;
        _aborted = false;
        return Result(best.Move, best.Value, best.Values, bestDepth, start, false, completed.ToArray());
    }

    private SearchResult Result(int move, double value, double?[] values, int depth, long start, bool det, int[] completed) =>
        new()
        {
            Move = move, Value = value, Values = values, Depth = depth,
            Nodes = _nodes, TTHits = _ttHits, TTSize = _ttSize,
            TimeUs = Timing.ElapsedUs(start), Deterministic = det, CompletedDepths = completed,
        };
}

/// <summary>Monotonic timing helpers.</summary>
public static class Timing
{
    public static long ElapsedNs(long start) =>
        (long)((Stopwatch.GetTimestamp() - start) * (1e9 / Stopwatch.Frequency));

    /// <summary>Elapsed microseconds, rounded half away from zero (Go math.Round).</summary>
    public static long ElapsedUs(long start) => (long)Math.Round(ElapsedNs(start) / 1000.0, MidpointRounding.AwayFromZero);

    public static double ElapsedMs(long start) => ElapsedNs(start) / 1e6;
}
